### PR TITLE
[SPARK-45697][BUILD] Fix `Unicode escapes in triple quoted strings are deprecated`

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to fix `Unicode escapes in triple quoted strings are deprecated` in scala 2.13.

### Why are the changes needed?
Eliminate compile warnings.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
- Pass GA.
- Manually test:
  ```
  ./build/sbt "catalyst/testOnly org.apache.spark.sql.catalyst.parser.PlanParserSuite"
  ```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -269,7 +269,8 @@ object SparkBuild extends PomBuild {
         // SPARK-45627 `enum`, `export` and `given` will become keywords in Scala 3,
         // so they are prohibited from being used as variable names in Scala 2.13 to
         // reduce the cost of migration in subsequent versions.
-        "-Wconf:cat=deprecation&msg=it will become a keyword in Scala 3:e"
+        "-Wconf:cat=deprecation&msg=it will become a keyword in Scala 3:e",
+        "-Wconf:cat=deprecation&msg=Unicode escapes in triple quoted strings are deprecated:e"
```
Comment (LuciferYang) [inline project/SparkBuild.scala:273]: this will be compile error in Scala 3?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 2]
```diff
@@ -1678,23 +1678,27 @@ class PlanParserSuite extends AnalysisTest {
           List.empty, List.empty, None, None, false)))
 
     // verify with ROW FORMAT DELIMETED
+    val collectionItemsTerminated1 = '\u0002'
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1681]: - '\u0002' corresponds to "STX" (Start of Text), a non-printing character used to mark the beginning of text.
- '\u0003' corresponds to "ETX" (End of Text), a non-printing character used to mark the end of text.
- '\u0004' corresponds to "EOT" (End of Transmission), a non-printing character used to mark the end of transmission.
- '\u0005' corresponds to "ENQ" (Enquiry), a non-printing character used to inquire whether the receiving device is ready to receive data.

I'm not sure if this cases really need to test using `non-printing characters` as delimiters, can we use printable characters for testing, so we can replace them with the actual corresponding characters. 

If indeed we need to test using `non-printing characters` as delimiters, I think we can add a `@nowarn` annotation to suppress warnings at a fine granularity like:

```
    // verify with ROW FORMAT DELIMITED
    @nowarn("cat=deprecation")
    val sqlWithRowFormatDelimited: String =
      """
        |SELECT TRANSFORM(a, b, c)
        |  ROW FORMAT DELIMITED
        |  FIELDS TERMINATED BY '\t'
        |  COLLECTION ITEMS TERMINATED BY '\u0002'
        |  MAP KEYS TERMINATED BY '\u0003'
        |  LINES TERMINATED BY '\n'
        |  NULL DEFINED AS 'null'
        |  USING 'cat' AS (a, b, c)
        |  ROW FORMAT DELIMITED
        |  FIELDS TERMINATED BY '\t'
        |  COLLECTION ITEMS TERMINATED BY '\u0004'
        |  MAP KEYS TERMINATED BY '\u0005'
        |  LINES TERMINATED BY '\n'
        |  NULL DEFINED AS 'NULL'
        |FROM testData
    """.stripMargin

    assertEqual(
      sqlWithRowFormatDelimited,
```

File: project/SparkBuild.scala (hunk 1 above)
Comment (panbingkun) [reply project/SparkBuild.scala:273]: Only for test in GA.
I want it to run through GA first and see if there are any other similar issues besides the above modifications.
I will delete it later.

**Comment by panbingkun** [comment]
image
image
image

File: project/SparkBuild.scala (hunk 1 above)
Comment (panbingkun) [reply project/SparkBuild.scala:273]: this will nothing in Scala 3.3.1, neither a warning nor an error.
- scala 3.3.1
   image

- scala 2.13.8
   image

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 3]
```diff
@@ -1677,25 +1677,24 @@ class PlanParserSuite extends AnalysisTest {
...
-        |  ROW FORMAT DELIMITED
-        |  FIELDS TERMINATED BY '\t'
-        |  COLLECTION ITEMS TERMINATED BY '\u0004'
-        |  MAP KEYS TERMINATED BY '\u0005'
-        |  LINES TERMINATED BY '\n'
-        |  NULL DEFINED AS 'NULL'
-        |FROM testData
-      """.stripMargin,
+    // verify with ROW FORMAT DELIMITED
+    val sqlWithRowFormatDelimited =
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1681]: looks a bit ugly ...

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 4]
```diff
@@ -1677,25 +1677,24 @@ class PlanParserSuite extends AnalysisTest {
...
+        "SELECT TRANSFORM(a, b, c)" + "\n" +
+        "  ROW FORMAT DELIMITED" + "\n" +
+      """  FIELDS TERMINATED BY '\t'""" + "\n" +
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1684]: friendly ping @AngersZhuuuu , do you have any suggestions about this `deprecated usage` issue

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala (hunk 3 above)
Comment (panbingkun) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1681]: image

**Comment by panbingkun** [comment]
image
image
image

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 5]
```diff
@@ -1677,25 +1677,23 @@ class PlanParserSuite extends AnalysisTest {
...
-        |  USING 'cat' AS (a, b, c)
-        |  ROW FORMAT DELIMITED
-        |  FIELDS TERMINATED BY '\t'
-        |  COLLECTION ITEMS TERMINATED BY '\u0004'
-        |  MAP KEYS TERMINATED BY '\u0005'
-        |  LINES TERMINATED BY '\n'
-        |  NULL DEFINED AS 'NULL'
-        |FROM testData
-      """.stripMargin,
+      "SELECT TRANSFORM(a, b, c)\n" +
```
Comment (panbingkun) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1682]: Although this writing may seem ugly, it seems better than the previous version, haha

**Review by dongjoon-hyun — COMMENTED** [review]
May I ask the status of this PR, @panbingkun ?

**Comment by panbingkun** [comment]
> May I ask the status of this PR, @panbingkun ?
@dongjoon-hyun 
My modification principle is to `eliminate the warning` and `maintain the original string semantics`, but it is indeed `ugly`;while @LuciferYang 's suggestion is to `eliminate the warning`, but bypass it with `a special annotation.`
Which `modification method` do you think is more appropriate? We would also like to hear your `suggestions`, many thanks!
(PS: I actually think both are acceptable, but a judgment and decision are needed. 😄)

**Comment by LuciferYang** [comment]
I am more inclined to understand the original intent of the test. If the goal is just to test different separators, we can use the literals of visible separators to eliminate this compile warning. If the original intent was to test invisible separators, then I think we can simply suppress the warning in this test.

**Comment by panbingkun** [comment]
> To clarify: I am more inclined to understand the original intent of the test. If the goal is just to test different separators, we can use the literals of visible separators to eliminate this compile warning. If the original intent was to test invisible separators, then I think we can simply suppress the warning in this test.

Okay, let's `suppress the warning` with annotations first.
I tested it locally and it was okay.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala [hunk 6]
```diff
@@ -1678,7 +1680,8 @@ class PlanParserSuite extends AnalysisTest {
           List.empty, List.empty, None, None, false)))
 
     // verify with ROW FORMAT DELIMETED
-    assertEqual(
+    @nowarn("cat=deprecation")
```
Comment (LuciferYang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/PlanParserSuite.scala:1683]: fine-grained suppressing this case with `@nowarn` if fine to me

File: pom.xml [hunk 7]
```diff
@@ -3000,6 +3000,7 @@
                 reduce the cost of migration in subsequent versions.
               -->
               <arg>-Wconf:cat=deprecation&amp;msg=it will become a keyword in Scala 3:e</arg>
+              <arg>-Wconf:cat=deprecation&amp;msg=Unicode escapes in triple quoted strings are deprecated:e</arg>
```
Comment (LuciferYang) [inline pom.xml:3003]: @panbingkun please revert the change of `SparkBuild.scala` and `pom.xml`

Comment (panbingkun) [reply pom.xml:3003]: Okay

**Review by LuciferYang — APPROVED** [review]

**Comment by LuciferYang** [comment]
Merged into master for Spark 4.0. Thanks @panbingkun
